python/tester/util.py

```diff
@@ -0,0 +1,184 @@
+
+    This number helps decide the next group number so it will not
+    collide with any previous groups.
+
+    Parameters
+    ----------
+    bucket : `s3.Bucket`
+        A S3 storage bucket
+    instrument : `str`
+        The short name of the active instrument.
+    date : `str`
+        The current date in YYYYMMDD format.
+
+    Returns
+    -------
+    group : `int`
+        The largest existing group for ``instrument``, or a newly generated
+        group if none exist.
+    """
+    preblobs = bucket.objects.filter(
+        Prefix=f"{instrument}/",
+    )
+    detector = min((int(preblob.key.split("/")[1]) for preblob in preblobs), default=0)
+
+    blobs = preblobs.filter(
+        Prefix=f"{instrument}/{detector}/{date}"
+    )
+    prefixes = [int(blob.key.split("/")[2]) for blob in blobs]
+    if len(prefixes) == 0:
+        return int(date) * 100_000
+    else:
+        return max(prefixes)
+
+
+def make_exposure_id(instrument, group_num, snap):
+    """Generate an exposure ID from an exposure's other metadata.
+
+    The exposure ID is designed for insertion into an image header, and is
+    therefore a string in the instrument's native format.
+
+    Parameters
+    ----------
+    instrument : `str`
+        The short name of the instrument.
+    group_num : `int`
+        The integer used to generate a group ID.
+    snap : `int`
+        A snap ID.
+
+    Returns
+    -------
+    exposure_key : `str`
+        The header key under which ``instrument`` stores the exposure ID.
+    exposure_header : `str`
+        An exposure ID that is likely to be unique for each combination of
+        ``group_num`` and ``snap``, for a given ``instrument``, in the format
+        for ``instrument``'s header.
+    exposure_num : `int`
+        An exposure ID equivalent to ``exposure_header`` in the format expected
+        by Gen 3 Middleware.
+    """
+    match instrument:
+        case "HSC":
+            return "EXP-ID", *make_hsc_id(group_num, snap)
+        case _:
+            raise NotImplementedError(f"Exposure ID generation not supported for {instrument}.")
+
+
+def make_hsc_id(group_num, snap):
+    """Generate an exposure ID that the Butler can parse as a valid HSC ID.
+
+    This function returns a value in the "HSCE########" format (introduced June
+    2016) for all exposures, even if the source image is older.
+
+    Parameters
+    ----------
+    group_num : `int`
+        The integer used to generate a group ID.
+    snap : `int`
+        A snap ID.
+
+    Returns
+    -------
+    exposure_header : `str`
+        An exposure ID that is likely to be unique for each combination of
+        ``group`` and ``snap``, in the form it appears in HSC headers.
+    exposure_num : `int`
+        The exposure ID genereated by Middleware from ``exposure_header``.
+
+    Notes
+    -----
+    The current implementation may overflow if more than ~60 calls to upload.py
+    are done on the same day.
```

Comment:
I rephrased it in https://github.com/lsst-dm/prompt_prototype/pull/40/commits/8624177d069b97cb1356651c3c8683f5313423cd. I think it reads a bit better now. 

Indeed once we scale up more and call `upload_hsc_rc2.py` with a larger group number, this will overflow. We might want a future ticket for that. I _think_ a longer string in the header should be fine if there aren't better ideas.

Reply:
Unfortunately, we really are limited to 8 digits: https://github.com/lsst/astro_metadata_translator/blob/main/python/astro_metadata_translator/translators/hsc.py#L221 will reject any longer string as invalid, and the old-style header format actually has fewer possible values. 😞

Reply:
I've filed [DM-37364](https://jira.lsstcorp.org/browse/DM-37364) for addressing the overflow issue.